MPTStates/MPT/MerklePatriciaNode.cs

```diff
@@ -0,0 +1,312 @@
+﻿using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text;
+using Neo.Cryptography;
+using Neo.IO;
+using Neo.Ledger;
+
+namespace Neo.Plugins.MPT
+{
+    /// <inheritdoc />
+    /// <summary>
+    /// Modified Merkel Patricia Node.
+    /// Note: It is not a thread safe implementation.
+    /// </summary>
+    public class MerklePatriciaNode : StateBase,
+        ICloneable<MerklePatriciaNode>,
+        IEquatable<MerklePatriciaNode>,
+        IEnumerable<byte[]>
+    {
+        private const int BranchSize = 18;
+        private const int ExtensionSize = 2;
+        private const int LeafSize = 3;
+
+        private byte[][] _hashes;
+
+        public MerklePatriciaNode() : this(0)
+        {
+        }
+
+        public MerklePatriciaNode(int size) => _hashes = new byte[size][];
+
+        /// <summary>
+        /// Indicates if the node is a branch.
+        /// </summary>
+        public bool IsBranch => _hashes.Length == BranchSize;
+
+        /// <summary>
+        /// Indicates if the node is an extension.
+        /// </summary>
+        public bool IsExtension => _hashes.Length == ExtensionSize;
+
+        /// <summary>
+        /// Indicates if the node is a leaf.
+        /// </summary>
+        public bool IsLeaf => _hashes.Length == LeafSize;
+
+        /// <summary>
+        /// Get and set the hashes by index.
+        /// </summary>
+        /// <param name="index">Index of the hash to get or set.</param>
+        public byte[] this[int index]
+        {
+            get => _hashes[index];
+            set => _hashes[index] = value;
+        }
+
+        /// <summary>
+        /// Get and set the path of the node.
+        /// Used for leaf and extension nodes.
+        /// </summary>
+        public byte[] Path
+        {
+            get => _hashes[0];
+            set => _hashes[0] = value;
+        }
+
+        /// <summary>
+        /// Get and set the key of the node.
+        /// Used for leaf and branch nodes.
+        /// </summary>
+        public byte[] Key
+        {
+            get => _hashes[_hashes.Length - 2];
+            set => _hashes[_hashes.Length - 2] = value;
```

Comment:
Throw an exception on empty tree

Reply:
I think node types are verified before usage.. IsNull for example, has no Key.